src/Assignments/SimpleProblemPage.tsx

```diff
@@ -6,98 +6,305 @@ import ProblemIframe from './ProblemIframe';
 import { BsCheckCircle, BsXCircle, BsSlashCircle } from 'react-icons/bs';
 import { ProblemDoneState } from '../Enums/AssignmentEnums';
 import _ from 'lodash';
-import { getQuestions } from '../APIInterfaces/BackendAPI/Requests/CourseRequests';
+import { endVersion, generateNewVersion, getQuestions, submitVersion } from '../APIInterfaces/BackendAPI/Requests/CourseRequests';
 import { ProblemDetails } from './ProblemDetails';
 import { ProblemStateProvider } from '../Contexts/CurrentProblemState';
+import { ConfirmationModalProps, ConfirmationModal } from '../Components/ConfirmationModal';
 
 interface SimpleProblemPageProps {
 }
 
 interface SimpleProblemPageLocationParams {
+    versionId?: string;
     topicId?: string;
     courseId?: string;
 }
 
 // This page has two panes. The left pane renders a list of questions, and the right pane renders the currently selected question.
 export const SimpleProblemPage: React.FC<SimpleProblemPageProps> = () => {
+    const DEFAULT_CONFIRMATION_PARAMETERS = {
+        show: false,
+        onConfirm: () => { console.error('onConfirm not set'); },
+        onHide: () => setConfirmationParameters(DEFAULT_CONFIRMATION_PARAMETERS),
+        headerContent: '',
+        bodyContent: ''
+    };
+
     const params = useParams<SimpleProblemPageLocationParams>();
     const [problems, setProblems] = useState<Record<number, ProblemObject> | null>(null);
     const [topic, setTopic] = useState<TopicObject | null>(null);
+    const [versionId, setVersionId] = useState<number | null>(null);
+    const [attemptsRemaining, setAttemptsRemaining] = useState<number>(1); // the only time these two are not set
+    const [versionsRemaining, setVersionsRemaining] = useState<number>(1); // is when an exam hasn't been attempted at all
+    const [modalLoading, setModalLoading] = useState<boolean>(false);
     const [selectedProblemId, setSelectedProblemId] = useState<number | null>(null);
     const [loading, setLoading] = useState(true);
     const [error, setError] = useState('');
+    const [confirmationParameters, setConfirmationParameters] = useState<ConfirmationModalProps>(DEFAULT_CONFIRMATION_PARAMETERS);
 
-    useEffect(()=>{
+    useEffect(() => {
         setLoading(true);
         (async () => {
             try {
-                if(_.isNil(params.topicId)) {
+                if (_.isNil(params.topicId)) {
                     console.error('topicId is null');
                     setError('An unexpected error has occurred');
                     return;
```

Comment:
If we return here, we never get out of the setLoading state, right? We should probably throw something instead.

Reply:
if we just
```TypeScript
throw new Error('An unexpected error has occurred');
```
then we don't even need to log it since it happens in the catch block

Reply:
updated -- though in the catch block we end up performing the same action 😂